proposals/NNNN-dxil18.md

```diff
@@ -0,0 +1,282 @@
+
+    // Record granularity (enum in 2 bits)
+    ThreadRecord = 0x20,   // [RW]ThreadNodeInputRecord, ThreadNodeOutputRecords
+    GroupRecord = 0x40,    // [RW]GroupNodeInputRecord, GroupNodeOutputRecords
+    DispatchRecord = 0x60, // [RW]DispatchNodeInputRecord
+    RecordGranularityMask = 0x60,
+
+    NodeIOKindMask = 0x7F,
+
+    TrackRWInputSharing = 0x100, // TrackRWInputSharing tracked on all non-empty
+                                 // input/output record/node types
+
+    // Mask for node/record properties beyond NodeIOKind
+    RecordFlagsMask = 0x100,
+    NodeFlagsMask = 0x100,
+  };
+
+  enum class NodeIOKind : uint32_t {
+    Invalid = 0,
+
+    EmptyInput =
+        (uint32_t)NodeIOFlags::EmptyRecord | (uint32_t)NodeIOFlags::Input,
+    NodeOutput =
+        (uint32_t)NodeIOFlags::ReadWrite | (uint32_t)NodeIOFlags::Output,
+    NodeOutputArray = (uint32_t)NodeIOFlags::ReadWrite |
+                      (uint32_t)NodeIOFlags::Output |
+                      (uint32_t)NodeIOFlags::NodeArray,
+    EmptyOutput =
+        (uint32_t)NodeIOFlags::EmptyRecord | (uint32_t)NodeIOFlags::Output,
+    EmptyOutputArray = (uint32_t)NodeIOFlags::EmptyRecord |
+                       (uint32_t)NodeIOFlags::Output |
+                       (uint32_t)NodeIOFlags::NodeArray,
+
+    DispatchNodeInputRecord =
+        (uint32_t)NodeIOFlags::Input | (uint32_t)NodeIOFlags::DispatchRecord,
+    GroupNodeInputRecords =
+        (uint32_t)NodeIOFlags::Input | (uint32_t)NodeIOFlags::GroupRecord,
+    ThreadNodeInputRecord =
+        (uint32_t)NodeIOFlags::Input | (uint32_t)NodeIOFlags::ThreadRecord,
+
+    RWDispatchNodeInputRecord = (uint32_t)NodeIOFlags::ReadWrite |
+                                (uint32_t)NodeIOFlags::Input |
+                                (uint32_t)NodeIOFlags::DispatchRecord,
+    RWGroupNodeInputRecords = (uint32_t)NodeIOFlags::ReadWrite |
+                              (uint32_t)NodeIOFlags::Input |
+                              (uint32_t)NodeIOFlags::GroupRecord,
+    RWThreadNodeInputRecord = (uint32_t)NodeIOFlags::ReadWrite |
+                              (uint32_t)NodeIOFlags::Input |
+                              (uint32_t)NodeIOFlags::ThreadRecord,
+
+    GroupNodeOutputRecords = (uint32_t)NodeIOFlags::ReadWrite |
+                             (uint32_t)NodeIOFlags::Output |
+                             (uint32_t)NodeIOFlags::GroupRecord,
+    ThreadNodeOutputRecords = (uint32_t)NodeIOFlags::ReadWrite |
+                              (uint32_t)NodeIOFlags::Output |
+                              (uint32_t)NodeIOFlags::ThreadRecord,
+  };
+```
+
+### Reading and Writing Node Records
+
+Unlike memory operations in earlier versions of DXIL, reading and writing node
+record memory uses LLVM's load and store instructions. DXIL 1.8 reserves address
+space 6 for node memory. The `getNodeRecordPtr` DXIL operation returns a pointer
+to node record memory in address space 6.

```

Comment:
I see now this addresses the space. I feel like maybe this would be more useful higher anyway and it would make the appearance of address space 6 in that table less surprising as well as the absence of some expected intrinsics.